Accordion

aiokaizen opened the pull request:
HTML Accordion support on FOAM

Commit:
Add Accordion view

Commit:
Fix pom issue

Commit:
Improvements to Accordion

Commit:
Update Accordion

Commit:
Merge branch 'development' into accordion

Commit:
Add hoverColor and tooltip support for Actions

Commit:
Merge branch 'development' into accordion

Commit:
Merge branch 'development' into accordion

Commit:
Merge branch 'development' into accordion

kgrgreer commented on src/foam/u2/Accordion.js:
```diff
@@ -0,0 +1,246 @@
+          this.start(foam.u2.Accordion, {
+            title: 'Action Accordion',
+            actions: [
+              {
```
Nobody would create an Action this way (that I've ever seen). They would define in their model's actions: section.

aiokaizen replied:
My usecase included context aware actions e.g. Display a different label based on the current scheme, or another context related value. I still have the actual action declared on the view / model, but in here I can add extra context arguments to it. If this already exists, please direct me on how to implement that instead.

kgrgreer replied:
Why would the label or tooltip be different in this use of the Action than it the others?
If you wanted to override particular values you could do SomeModel.SOME_ACTION.clone().copyFrom({label: 'custom'})

aiokaizen replied:
I agree, I was looking at some examples from the repo and I understand what you say, however, if you can just approve it now in order to check if other stuff on PTV3 are working properly, after that, I'll fix this and recreate the PR with the new method.

sarthak-marwaha replied:
If you just need an array of Actions then this should be an FObjectArray, if you care about the data linked to that action and want to specify it per action, have a look at ButtonGroup, it does something very similar, in fact you can replace all this logic with a single ButtonGroup

aiokaizen replied:
I had changed the entire layout and removed this functionality all together. Now it is just an actions section where the user can put what ever they want. They can use it to display not just actions, but also counters, icons, etc...

kgrgreer commented on src/foam/u2/Accordion.js:
```diff
@@ -0,0 +1,246 @@
+      const self = this;
+
+      let toolbar = this
+        .addClass(this.myClass())
```
.addClass() is the same but shorter

kgrgreer commented on src/foam/u2/Accordion.js:
```diff
@@ -0,0 +1,246 @@
+        .enableClass('expanded', this.expanded$)
+        .start('div')
+          .addClass(this.myClass('toolbar'))
+          .on('click', _ => this.toggle());
```
toggle should be a listener and then you would just need to do: .on('click', this.toggle);

kgrgreer replied:
Or if you need both an Action and listener have toggle action and onToggle listener and have action call listener.

aiokaizen replied:
Yes that's exactly my case. I have added onToggle.

kgrgreer commented on src/foam/u2/Accordion.js:
```diff
@@ -0,0 +1,246 @@
+          .addClass(this.myClass('title'));
+
+      if ( this.expandIconPosition === 'left' ) {
+        titleSection.tag(
```
Why not just .add(this.TOGGLE) directly and configure the icon in the action:?

aiokaizen replied:
Didn't know I can do that, I have changed it as well. Got some slight change in the ui but that's fine, I'll fix it as well.

kgrgreer replied:
You can add properties and actions directly: like this.Person.USER_NAME and this.Email.SEND and will add the appropriate view and bind it to the data properly with validation, visibility, i18n, accessibility, etc.
Just adding a single button when you need it to be i18n, a11y, themed, permissioned, mode aware, debounced, is/available/enabled aware, is too much work to do manually. Same for property views.

aiokaizen replied:
Understood. I've already changed this, I will look for other instances in PTV3 as well.

kgrgreer commented on src/foam/u2/Accordion.js:
```diff
@@ -0,0 +1,246 @@
+        ).addClass(this.myClass('toggle'))
+      }
+
+      if ( this.icon !== undefined ) {
```
You don't need to make titleSection a variable. Then code like this could have been chained as:
this.callIf(this.icon(), function () { this.tag(this.icon); })....

aiokaizen replied:
Got it, I got rid of all the variables in my render function, now it's just a single flow.

kgrgreer commented on src/foam/u2/Accordion.js:
```diff
@@ -0,0 +1,246 @@
+        .start()
+          .addClass(this.myClass('actions'));
+
+      if ( this.actions.length > 0 ) {
```
Why is this if necessary?

aiokaizen replied:
You're right, it's not. I use forEach now.

kgrgreer commented on src/foam/u2/Accordion.js:
```diff
@@ -0,0 +1,246 @@
+      if ( this.actions.length > 0 ) {
+        for ( let action of this.actions ) {
+          const actionId = `${this.id}-${action.name}`;
+          let actionEl = actionsSection.start(self.ActionView, {
```
Just add the action directly.

kgrgreer commented on src/foam/u2/Accordion.js:
```diff
@@ -0,0 +1,246 @@
+          if ( action.name ) {
+            actionEl.addClass(self.myClass(action.name));
+          }
+          if ( action.tooltip ) {
```
Just add to the action

kgrgreer commented on src/foam/u2/Accordion.js:
```diff
@@ -0,0 +1,246 @@
+          if ( action.tooltip ) {
+            actionEl.tooltip = action.tooltip;
+          }
+          if ( action.hoverColor ) {
```
Why is this necessary?

aiokaizen replied:
changed to `callIf`, it provides a custom color on hover.

kgrgreer commented on src/foam/u2/Accordion.js:
```diff
@@ -0,0 +1,246 @@
+                label: 'Action 1',
+                tooltip: 'Action 1 Tip.',
+                icon: 'images/my-action-icon.svg',
+                hoverColor: 'red',
```
Is hoverColor actually a thing?

aiokaizen replied:
No, I implemented it manually in the Accordion.

sarthak-marwaha replied:
Is this really something that needs to be provided per instance of this view? Could it not be derived from the color of the background? See https://github.com/kgrgreer/foam3/blob/848c335316bd3842d43dd9f5ab17a8863144d38a/src/foam/u2/tag/Button.js#L123-L125

aiokaizen replied:
You're right, the new layout removes the actions all together, including this part. Now the user is responsible for providing their own actions.

Commit:
Fix PR comments

kgrgreer commented on src/foam/u2/Accordion.js:
```diff
@@ -0,0 +1,247 @@
+              this.start(self.ActionView, {
+                id: actionId,
+                action: action.action,
+                data: action.data || self,
```
Do actions have data?

sarthak-marwaha replied:
What you are looking for are called ActionReference, they can store an Action and data that the action should be associated with. Then you can simply call .start(youActionReference)

kgrgreer commented on src/foam/u2/Accordion.js:
```diff
@@ -0,0 +1,247 @@
+            .forEach(this.actions, function(action) {
+              const actionId = `${self.id}-${action.name}`;
+              this.start(self.ActionView, {
+                id: actionId,
```
Why set the id?

aiokaizen replied:
the Element id that is rendered in html e.g. <div *id="some-id"*>.

kgrgreer commented on src/foam/u2/Accordion.js:
```diff
@@ -0,0 +1,247 @@
+              const actionId = `${self.id}-${action.name}`;
+              this.start(self.ActionView, {
+                id: actionId,
+                action: action.action,
```
What is action? What is action.action?

aiokaizen replied:
action is an item of the `actions` list. `action.action` is the action itself.

kgrgreer commented on src/foam/u2/Accordion.js:
```diff
@@ -0,0 +1,247 @@
+                id: actionId,
+                action: action.action,
+                data: action.data || self,
+                icon: action.icon || '',
```
Why do you need to set this?

aiokaizen replied:
When I'm finished, it would probably just be a list of actions. e.g. 
```
   actions: [ actions1, action2 ]
```

sarthak-marwaha replied:
You can loop over those by wrapping this in a forEach

aiokaizen replied:
Removed in the new layout.

kgrgreer commented on src/foam/u2/Accordion.js:
```diff
@@ -0,0 +1,247 @@
+                action: action.action,
+                data: action.data || self,
+                icon: action.icon || '',
+                label: action.label
```
Again, why set? ActionView will just take it from the action itself?

aiokaizen replied:
When I'm finished, it would probably just be a list of actions. e.g. 
```
   actions: [ actions1, action2 ]
```

kgrgreer commented on src/foam/u2/Accordion.js:
```diff
@@ -0,0 +1,247 @@
+                label: action.label
+              })
+              .addClass(self.myClass('action'))
+              .callIf(action.name, function() {
```
Doesn't action view already give it custom CSS name based on the action's name?

aiokaizen replied:
It actually does. That's redundant from my part. I'll remove it.

kgrgreer commented on src/foam/u2/Accordion.js:
```diff
@@ -0,0 +1,247 @@
+              .callIf(action.name, function() {
+                this.addClass(self.myClass(action.name));
+              })
+              .callIf(action.tooltip, function() {
```
Same.

sarthak-marwaha commented on src/foam/u2/Accordion.js:
```diff
@@ -0,0 +1,247 @@
+    {
+      name: 'toggle',
+      label: '',
+      icon: 'images/cheveron-right.svg',
```
Can use:
```suggestion
      themeIcon: 'next',
```
It will render svgs instead of images. Much better quality and scaling

kgrgreer commented on src/foam/u2/Accordion.js:
```diff
@@ -0,0 +1,247 @@
+              .callIf(action.tooltip, function() {
+                this.tooltip = action.tooltip;
+              })
+              .callIf(action.hoverColor, function() {
```
If someone wants to do this they can just add the appropriate :hover code to their css:.

aiokaizen replied:
Correct, I have removed this feature.

sarthak-marwaha commented:
@aiokaizen Can we get some screenshots to see how this looks?

kgrgreer commented on src/foam/u2/Accordion.js:
```diff
@@ -0,0 +1,247 @@
+      `,
+    },
+    {
+      name: 'actions',
```
Are these Actions or ActionViews?

aiokaizen replied:
Actions

kgrgreer commented on src/foam/u2/Accordion.js:
```diff
@@ -0,0 +1,247 @@
+    },
+    {
+      name: 'actions',
+      value: [],
```
Should never do value: [] otherwise all instances will share the same array (unless they set to something else).
Instead do:
factory: function() { return []; }
and then each one will get their own array.
Or, make the class: 'Array' and specify the of: and then it will set the factory for you.

kgrgreer commented on src/foam/u2/Accordion.js:
```diff
@@ -0,0 +1,247 @@
+            .addClass(this.myClass('title'))
+            .callIf(this.expandIconPosition === 'left', function() {
+              this.start(
+                self.TOGGLE
```
Because this is an Action, you don't need the click listener. The ActionView will call it for you.

aiokaizen replied:
I have removed the onClick from the buttons.

kgrgreer commented on src/foam/u2/Accordion.js:
```diff
@@ -0,0 +1,247 @@
+    }
+  ],
+
+  listeners: [
```
Not needed because the ActionView will automatically call toggle() anyway.

aiokaizen replied:
I need it for when the user clicks on the toolbar itself, not just the toggler.

sarthak-marwaha replied:
You can just call this.toggle() from anywhere in your code and it will call the action

Commit:
Merge branch 'development' into accordion

aiokaizen commented:
@sarthak-marwaha 

![image](https://github.com/user-attachments/assets/1f4d00a1-115b-4003-a968-ffd4b9b47e4c)

![image](https://github.com/user-attachments/assets/cf7ce4d1-afc1-4aef-94b0-806a3104c525)

sarthak-marwaha commented on src/foam/u2/Accordion.js:
```diff
@@ -0,0 +1,247 @@
+    ^ {
+      width: 100%;
+      overflow: hidden;
+      background-color: #eee;
```
This is not ideal for styling foam elements, this should point to a CSSToken, the default ones are available in CSSTokens.js. You could find the closest match to you and replace this with that, then in your deployment, you can override them with an exact match if required. Same for all color values.

aiokaizen replied:
I have changed all my colors to use CSSTokens.

sarthak-marwaha commented on src/foam/u2/Accordion.js:
```diff
@@ -0,0 +1,247 @@
+      border: 1px solid #eee;
+      border-bottom-color: #ccc;
+    }
+    ^:first-child {
```
Why not just add padding to to ^?

aiokaizen replied:
The accordion is usually used in groups, this helps in providing a nice group of accordions, while separating the first and last children from other elements in the page.

sarthak-marwaha replied:
In that case shouldnt the view that is grouping the accordions? This way you are forcing styling on every use case. And everyone will have to use it in the same way. FOAM views are written to fill the space they are provided. They should not assume any styling the parent view might want, this is what makes them reusable everywhere.

aiokaizen replied:
I agree. I will remove the margins from Accordion and add them to the parents from my code.

sarthak-marwaha commented on src/foam/u2/Accordion.js:
```diff
@@ -0,0 +1,247 @@
+      margin-top: 15px;
+      border-radius: 5px 5px 0 0;
+    }
+    ^:last-child {
```
Same here

kgrgreer commented:
Better to remove Action support altogether and just exponse three content areas: content, title, toolbar and then let users add whatever they want to those areas.

Commit:
Use theme colors instead of hard-coded ones.

Commit:
Merge branch 'development' into accordion

sarthak-marwaha commented on src/foam/u2/Accordion.js:
```diff
@@ -47,11 +47,12 @@ foam.CLASS({
     ^action:hover {
-      background-color: rgb(74, 170, 135) !important;
-      color: white !important;
+      background-color: $surface-primary-normal!important;
```
Where does this cssToken come from? if this is only in your repo, set a standard color here and then override it in your journal

aiokaizen replied:
You're correct, the equivalent is `$primary500`, but it's removed anyways.

Commit:
Update layout:
  - Change actions behaviour from litteral actions to action section where
    The user can put whatever they want. Less API overhead, more generic.
  - Allow the title section to accept any element, allowing the user to
    insert icons, text, counters, ...
  - Add a toolbar section that allows the user to override the entire toolbar
    section allowing for even more generic usecases.

aiokaizen committed:
Update src/foam/u2/Accordion.js

Replace icon with a themeIcon for better quality and scaling

Co-authored-by: sarthak <81172969+sarthak-marwaha@users.noreply.github.com>

Commit:
Minor changes

kgrgreer commented on src/foam/u2/Accordion.js:
```diff
@@ -0,0 +1,176 @@
+        .addClass()
+        .enableClass('expanded', this.expanded$)
+        .callIfElse(this.toolbar, function() {
+          this.start('div')
```
This line and the next line are the same for both the 'if' and 'else' portions, so could be moved outside the callIfElse().

kgrgreer commented on src/foam/u2/Accordion.js:
```diff
@@ -0,0 +1,176 @@
+                  });
+          })
+
+      this
```
Are these three lines duplicated above?

aiokaizen replied:
Yes they were, good catch. I was moving code around and didn't notice them.

sarthak-marwaha commented on src/foam/u2/Accordion.js:
```diff
@@ -0,0 +1,176 @@
+    }
+    ^toolbar {
+      background-color: $grey50;
+      padding: 5px;
```
FOAM follows a 4px grid, can this be 4px or 8px instead?

sarthak-marwaha commented on src/foam/u2/Accordion.js:
```diff
@@ -0,0 +1,176 @@
+      margin: 0;
+    }
+    ^title {
+      padding: 3px;
```
Same here

sarthak-marwaha commented on src/foam/u2/Accordion.js:
```diff
@@ -0,0 +1,176 @@
+      opacity: 1;
+    }
+    ^ ^toggle svg {
+      width: 12px;
```
Do you need this? Can you instead set `size: 'SMALL'` on the action? It should do the same thing

aiokaizen replied:
It generates a width / height of 1.15rem which ends up larger than my requirement.

sarthak-marwaha replied:
Okay can you replace this with rem then so this scales properly with font sizing. FOAM uses a base 1rem = 10px so this would be width: 1.2rem. You can see Fonts.js where it updates the font-size of :root. If your 1.15rem is larger then you are overriding that value somewhere

aiokaizen replied:
Sure

sarthak-marwaha commented on src/foam/u2/Accordion.js:
```diff
@@ -0,0 +1,176 @@
+        .callIfElse(this.toolbar, function() {
+          this.start('div')
+            .addClass(self.myClass('toolbar'))
+            .on('click', this.onToggle)
```
```suggestion
            .on('click', self.toggle.bind(self))
```

sarthak-marwaha replied:
You can remove the listener after you do this

aiokaizen replied:
Outdated :/ however, I committed it from my end. Thanks for the optimization.

sarthak-marwaha commented:
I hate to do this now but do either of these meet your requirements? @aiokaizen @kgrgreer https://github.com/kgrgreer/foam3/blob/6ed83dc83adeac098f4d518a68dd2986d66acc33/src/foam/u2/borders/ExpandableBorder.js

https://github.com/kgrgreer/foam3/blob/6ed83dc83adeac098f4d518a68dd2986d66acc33/src/foam/u2/borders/CollapseBorder.js

Commit:
Fix some github comments

Commit:
Follow foam's 4x4 grid and fix a small bug in the right toggler

Commit:
Remove the listner and replace it with toggle.bind

aiokaizen commented:
@sarthak-marwaha I think the closest thing to what I'm doing now is `CollapseBorder.js` https://github.com/kgrgreer/foam3/blob/6ed83dc83adeac098f4d518a68dd2986d66acc33/src/foam/u2/borders/CollapseBorder.js

But it lacks the following functionality:
- flexible toggler positioning (right or left)
- actions section

We could update CollapseBorder but the updates could break old codebase. Since the Accordion is more generic and flexible, and  the name is clearer and already a standard as an HTML component used everywhere; I suggest deprecating both ExpandableBorder and CollapseBorder in favour of Accordion. This would result in:
- Lesser code, since both ExpandableBorder and CollapseBorder would be replaced by a single component.
- More flexibility, Accordion is more flexible than both, while providing all their functionalities.
- Clearer component. You don't need to explain what an Accordion is to a developer, where is you need to do so for both ExpandableBorder and CollapsableBorder.

@kgrgreer what is your oppinion on this?

Commit:
Remove top and bottom margins

Commit:
Merge branch 'development' into accordion

sarthak-marwaha commented on src/foam/u2/images/triangle.svg:
Is this used anywhere?

aiokaizen replied:
Was used in the Accordion, let me remove it real quick

sarthak-marwaha commented on src/foam/u2/images/cheveron-right.svg:
Can be deleted

aiokaizen replied:
True, now that we use themeIcon, it's no longer needed.

Commit:
Remove unused SVGs

sarthak-marwaha commented on src/foam/u2/Accordion.js:
```diff
@@ -0,0 +1,160 @@
+    }
+    ^ .foam-u2-ActionView-toggle {
+      transition: transform 0.3s;
+      background: transparent;
+      border-radius: 50%;
+      border: none;
+      outline: none;
+      outline: none;
+      padding: 4px;
```
Setting buttonStyle: 'UNSTYLED' on the action should handle this

aiokaizen replied:
Very handy. Thanks

sarthak-marwaha commented on src/foam/u2/Accordion.js:
```diff
@@ -0,0 +1,160 @@
+      height: 28px;
+    }
+    ^ .foam-u2-ActionView-toggle:hover {
+      background: transparent !important;
```
Unstyled will fix this too

Commit:
Replace px with rem

sarthak-marwaha commented on src/foam/u2/Accordion.js:
```diff
@@ -0,0 +1,160 @@
+      outline: none;
+      outline: none;
+      padding: 4px;
+      width: 28px;
+      min-width: 28px;
+      height: 28px;
```
replace with rem

sarthak-marwaha commented on src/foam/u2/Accordion.js:
```diff
@@ -0,0 +1,160 @@
+
+  properties: [
+    'title',
+    'actions',
+    'toolbar',
```
Can we get some documentation for these

sarthak-marwaha commented:
> @sarthak-marwaha I think the closest thing to what I'm doing now is `CollapseBorder.js` https://github.com/kgrgreer/foam3/blob/6ed83dc83adeac098f4d518a68dd2986d66acc33/src/foam/u2/borders/CollapseBorder.js
> 
> But it lacks the following functionality:
> 
> * flexible toggler positioning (right or left)
> * actions section
> 
> We could update CollapseBorder but the updates could break old codebase. Since the Accordion is more generic and flexible, and the name is clearer and already a standard as an HTML component used everywhere; I suggest deprecating both ExpandableBorder and CollapseBorder in favour of Accordion. This would result in:
> 
> * Lesser code, since both ExpandableBorder and CollapseBorder would be replaced by a single component.
> * More flexibility, Accordion is more flexible than both, while providing all their functionalities.
> * Clearer component. You don't need to explain what an Accordion is to a developer, where is you need to do so for both ExpandableBorder and CollapsableBorder.
> 
> @kgrgreer what is your oppinion on this?

I would like to keep at least ExpandableBorder, since it is detached from all controls and offers unique functionality. I am okay with deprecating CollapseBorder eventually. Added some more clean up comments on this

sarthak-marwaha commented on src/foam/u2/Accordion.js:
```diff
@@ -0,0 +1,160 @@
+  name: 'Accordion',
+  extends: 'foam.u2.Controller',
+
+  requires: [ 'foam.u2.ActionView' ],
```
remove

kgrgreer commented on src/foam/u2/Accordion.js:
```diff
@@ -0,0 +1,160 @@
+
+  methods: [
+    function init() {
+
```
Remove blank line

kgrgreer commented on src/foam/u2/Accordion.js:
```diff
@@ -0,0 +1,160 @@
+                  this.start(self.TOGGLE)
+                    .addClass(self.myClass('toggle'))
+                });
+          })
```
Missing ;

kgrgreer commented on src/foam/u2/Accordion.js:
```diff
@@ -0,0 +1,160 @@
+      background: transparent;
+      border-radius: 50%;
+      border: none;
+      outline: none;
```
This line is duplicated.

kgrgreer commented:
> I hate to do this now but do either of these meet your requirements? @aiokaizen @kgrgreer https://github.com/kgrgreer/foam3/blob/6ed83dc83adeac098f4d518a68dd2986d66acc33/src/foam/u2/borders/ExpandableBorder.js
> 
> https://github.com/kgrgreer/foam3/blob/6ed83dc83adeac098f4d518a68dd2986d66acc33/src/foam/u2/borders/CollapseBorder.js

We could add the 'actions' section to CollapseBorder and rename it to Accordion, since it is only used in one place in FOAM (I don't know about NP). If people don't use the 'actions' section, it wouldn't hurt.

Commit:
Add documentation and fix comments

sarthak-marwaha commented:
> > I hate to do this now but do either of these meet your requirements? @aiokaizen @kgrgreer https://github.com/kgrgreer/foam3/blob/6ed83dc83adeac098f4d518a68dd2986d66acc33/src/foam/u2/borders/ExpandableBorder.js
> > https://github.com/kgrgreer/foam3/blob/6ed83dc83adeac098f4d518a68dd2986d66acc33/src/foam/u2/borders/CollapseBorder.js
> 
> We could add the 'actions' section to CollapseBorder and rename it to Accordion, since it is only used in one place in FOAM (I don't know about NP). If people don't use the 'actions' section, it wouldn't hurt.

Happy with this solution.

aiokaizen commented:
Fair enough. This means that I shall:
1. Implement actions, and toolbar sections in **CollapseBorder**
2. Rename **CollapseBorder** to **Accordion**
3. Remove my own Accordion class.
4. Change all occurrences of **CollapseBorder** to use the new name.
5. I will keep the old styling for CollapseBorder and implement my accordion styles from my end in PTV3.

Question: **CollapseBorder** lives in `foam.u2.borders`, while **Accordion** is located in `foam.u2`. I believe the latter is more suitable for the new name. I'll do it like so, if you think it's the other way around, I'll change it when I get your feedback.

Commit:
Rename foam.u2.borders.CollapseBorder to foam.u2.Accordion

aiokaizen commented:
I ended up Refactoring the foam.u2.borders.CollapseBorder to foam.u2.Accordion. I preserved CollapseBorder design for compatibility with old code. I also cleaned up some unused CSS rules from CollapseBorder side and kept only the relevant ones.

Commit:
Remove unused require

Commit:
Minor css class change

sarthak-marwaha commented on src/foam/u2/Accordion.js:
```diff
@@ -0,0 +1,184 @@
+        USAGE:
+          this.tag(foam.u2.Accordion, {
+            title: "User"
+            actions: foam.u2.Element.create()
```
@kgrgreer is it safe to add elements like this now? We preferred viewspecs in u2 but does it not matter in u3?

kgrgreer replied:
It should be fine in U3 except that the way he's done it it will have the wrong Context. I'm not sure how those actions are working. Oh, I see, it's just a document. If he imported Element then he could do this.Element.create() which would work but the better approach would be to not specify actions: ahead of time but instead do:
this.start(foam.u2.Accordion).call(function() {
  this.actions.add(self.UPDATE_ACTION, self.DELETE_ACTION));
});
This way they will get the sub-context of the Accordion, which is correct, and it's actually less code.

sarthak-marwaha replied:
Or just add a dsl addActions method. Then it works even better.

this.start(foam.u2.Accordion).addActions(self.UPDATE_ACTION, self.DELETE_ACTION));

aiokaizen replied:
@sarthak-marwaha I have implemented @kgrgreer 's proposition, since I don't want to force explicit actions to be added, the user can add what ever they want to the actions component. Also, I think **`actions`** is misleading so, I'm changing the name to something more meaningful. I'm choosing **`rightSlot`** since it is a slot that can hold any kind of elements. I'm open to any other suggestions.

sarthak-marwaha commented on src/foam/u2/Accordion.js:
```diff
@@ -0,0 +1,184 @@
+      flex-direction: row;
+      justify-content: space-between;
+      align-items: center;
+      left: 1.5vmin;
+      top: max(-12px, -2vmin);
```
What are these lines doing?

aiokaizen replied:
I don't actually know! I picked them from CollapseBorder code. Didn't want to remove them since they might have some impact elsewhere. Not even familiar with the syntax (1.5vmin).

sarthak-marwaha replied:
I think remove them, they look like they are adding padding but that can just be done with the padding property

Commit:
Remove unnecessary css rules.

kgrgreer commented on src/foam/u2/Accordion.js:
```diff
@@ -0,0 +1,182 @@
+
+        USAGE:
+          this.tag(foam.u2.Accordion, {
+            toolbar: foam.u2.Element.create()
```
See the above example.

Commit:
Update actions documentation and code.

Commit:
Update Accordion actions section

Commit:
Remove toolbar section and apply some comments.

Commit:
Replace  with a function since It was not working

Commit:
Merge branch 'development' into accordion
